corehq/apps/hqwebapp/tests/utils/test_translation.py

```diff
@@ -0,0 +1,78 @@
+from contextlib import ContextDecorator
+from unittest.mock import MagicMock, patch
+from django.test import SimpleTestCase
+
+from django.utils import translation
+from django.utils.translation import ugettext_lazy
+from django.utils.translation.trans_real import translation as get_translations
+
+from ...utils.translation import mark_safe_lazy, format_html_lazy
+
+CUSTOM_LANGUAGE = 'custom'
+
+
+class custom_translations(ContextDecorator):
+    """
+    A decorator/context manager to provide runtime translations for the 'custom' language
+    """
+    def __init__(self, translation_mapping):
+        self.translation_mapping = translation_mapping
+
+    def __enter__(self):
+        translations = get_translations(CUSTOM_LANGUAGE)
+        old_gettext = translations.gettext
+
+        def lookup(id):
+            return self.translation_mapping.get(id) or old_gettext(id)
+
+        self.patcher = patch.object(translations, 'gettext', MagicMock(side_effect=lookup))
+        self.patcher.start()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.patcher.stop()
+
+
+class TestLazyMarkSafe(SimpleTestCase):
+    @custom_translations({'Translate Me': 'Translated'})
+    def test_lazy_translation(self):
+        translation_promise = mark_safe_lazy(ugettext_lazy('Translate Me'))
+
+        with translation.override(CUSTOM_LANGUAGE):
+            translated = str(translation_promise)
+
+        self.assertEqual(translated, 'Translated')
```

Comment:
Would it make sense to add a test for the default case to verify that the `custom_translations` decorator is not somehow overriding all translation logic, even when the "custom" translation is not active?
```suggestion
        self.assertEqual(translated, 'Translated')

    @custom_translations({'Translate Me': 'Translated'})
    def test_default_translation(self):
        from django.utils.translation import get_language
        translation_promise = mark_safe_lazy(ugettext_lazy('Translate Me'))
        self.assertNotEqual(get_language(), CUSTOM)
        self.assertEqual(str(translation_promise), 'Translate Me')
```

Reply:
I was trying to use naming as best I could to handle that concern. If you feel it's still a concern, then I think the correct approach is making a test suite for the decorator itself -- I don't think the above test tests any behavior for `mark_safe_lazy`/`format_html_lazy`

Reply:
Sure, putting this in a test for the decorator would be fine. The reason I suggested it was because I was having a hard time following the logic in the decorator to figure out when the custom translation is active and when it's not. I eventually read the source code of `django.utils.translation` to get a better understanding of what was going on.